### DISCUSSION TITLE
How to achieve the constraint - "Material Availability", in the Food packaging example?

### DISCUSSION THREAD

**Discussion by kentbill** [TimefoldAI/timefold-solver#238] [General]
Hi there
In the Planning Mode - [Food packaging](https://timefold.ai/model/food-packaging/), there is a constraint "Material Availability" is quite similar to my scene:

> **Material Availability**: The availability of packaging materials like film, foil, cardboard, and labels can limit when certain jobs can be scheduled. For instance, if a specific type of film will only be available starting next week, jobs requiring that film cannot be scheduled for this week.

But I can't find the implementation of this constraint in the code, has it been implemented yet?

In my scenario, there is a slight difference from this constraint. I define materials as non-renewable resources (i.e. non-renewable resources in the Project Job Scheduling example), and then the daily available amount of resources is the sum of the remaining amount from the previous day and the received amount. Today's tasks can only begin when the available resources are greater than or equal to today's task requirements.
So how to achieve such this case in ConstraintStream?

Many thanks

**Comment by nicksi** [comment]
Look like it is not implemented yet. The direct (and probably suboptimal) approach will be to create a listener? triggered by job assignment, which will recalculate daily resource reminders. After that you can penalise for each day with resources below zero.

**Reply by kentbill** [reply]
Hmm, Thank you Nikolai. Maybe using a ShadowVariable to determine the start time of the task is one of the good ways, but I haven't tried it yet. I'm trying to use the resource capacity limit constraint to achieve it, that is, if the amount of resources on a certain day is the sum of the remaining resources of the previous day and the amount of resources received in the warehouse (that is, the amount of storage) that day, which is not enough for one task, it will be calculated as The amount of missing resources acts as a penalty score.

**Comment by ge0ffrey** [comment]
In the food packaging quickstarts, there's Job.readyDateTime and dueDateTime to limit the time window of a job.
https://github.com/TimefoldAI/timefold-quickstarts/blob/stable/use-cases/food-packaging/src/main/java/org/acme/foodpackaging/solver/FoodPackagingConstraintProvider.java#L18
Note that readyTime is handled through the VariableListener (automatic waiting).

In the project job scheduling example, there is the concept of renewable and nonRenawable resources:
  https://github.com/TimefoldAI/timefold-solver/blob/main/examples/src/main/java/ai/timefold/solver/examples/projectjobscheduling/score/ProjectJobSchedulingConstraintProvider.java#L18
but note that that example needs to cleaning up.

I advise to start from the food packaging quickstart and look at the project job scheduling for inspiration.

**Reply by ge0ffrey** [reply]
but doing it with shadow vars might indeed be better, as nicksi suggested.

**Comment by kentbill** [comment]
Thank you, Geoffrey.

I know that in the FPS example, you can use job.readyDateTime to determine the start time of a job, and then set the material's arrival time as job.readyDateTime. 

But my scenario may be a little more complicated. The arrival time of the material is not the condition for the start time of the job, but the 
 material's quantity. That is to say, when a job gets a start time, if the quantity of materials it requires at this time is greater than its requirement, the start time is feasible, otherwise it may have to wait for the arrival of the next batch of materials, or there is another job occupied this material, is scheduled to another time, so that the job gets enough material, its start time is feasible.

So, my method is to add a method for Material (which I define the material as a type of resource), which can calculate the daily available quantity of materials.

The material's available quantity for a day is the sum of the remaining quantity from the previous day and the received quantity for that day. Then a capacity constraint similar to resources in the PJS sample which used to ensure that the sum of material requirements for jobs on that day is less than the available quantity.
